kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -93,15 +95,32 @@ public CloseableIterator<FilteredColumnarBatch> getScanFiles(TableClient tableCl
 
     @Override
     public Row getScanState(TableClient tableClient) {
+        // Physical equivalent of the logical read schema.
+        StructType physicalReadSchema = ColumnMapping.convertToPhysicalSchema(
+            readSchema,
+            snapshotSchema,
+            ColumnMapping.getColumnMappingMode(metadata.getConfiguration()));
+
+        // Compute the physical data read schema, basically the list of columns to read
+        // from a Parquet data file. It should exclude partition columns and include
+        // row_index metadata columns (in case DVs are present)
+        List<String> partitionColumns = VectorUtils.toJavaList(metadata.getPartitionColumns());
+        StructType physicalDataReadSchema =
+            PartitionUtils.physicalSchemaWithoutPartitionColumns(
+                readSchema, /* logical read schema */
+                physicalReadSchema,
+                new HashSet<>(partitionColumns)
+            )
+            // TODO: do we only want to request row_index_col when there are DVs?
+            // Is there a way to identify this with table property?
+            .add(StructField.METADATA_ROW_INDEX_COLUMN);
+
         return ScanStateRow.of(
             metadata,
             protocol,
             readSchema.toJson(),
-            ColumnMapping.convertToPhysicalSchema(
-                readSchema,
-                snapshotSchema,
-                ColumnMapping.getColumnMappingMode(metadata.getConfiguration())
-            ).toJson(),
+            physicalReadSchema.toJson(),
```

Comment:
We need two physical schemas
1) physical read schema (contains partition and non-partition columns) - need physical column names for partitions columns when looking up partition value in `AddFile`.
2) readPhysicalDataSchema (contains data columns and metadata columns) - this is the schema for reading data from Parquet files

I was trying to avoid computing readPhysicalDataSchema multiple times, but actually, it increases the scan state row. 

Couple of approaches:
1) Moving the computation of physical data read schema to ScanStateRow as a utility method to derive the physical schema. 
2) Scan state contains the partition column map of logical name to physical name and the physical data read schema (computed once at the driver).

Reply:
So if I understand correctly, we have:
* logical read schema (= what the query asked for)
* physical read schema (= applies column mapping)
* data read schema (= subset of physical read schema that actually goes to the parquet reader)

And then we work backward during execution:
* read data from parquet
* apply any missing columns to make it match the physical read schema
* rename as needed to match logical read schema

Is that right?